vault/resource_kubernetes_secret_backend_role.go

```diff
@@ -0,0 +1,257 @@
+package vault
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"regexp"
+	"strings"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/hashicorp/terraform-provider-vault/internal/consts"
+	"github.com/hashicorp/terraform-provider-vault/internal/provider"
+)
+
+var (
+	kubernetesSecretBackendFromPathRegex = regexp.MustCompile("^(.+)/roles/.+$")
+)
+
+const (
+	fieldAllowedKubernetesNamespaces = "allowed_kubernetes_namespaces"
+	fieldTokenMaxTTL                 = "token_max_ttl"
+	fieldTokenDefaultTTL             = "token_default_ttl"
+	fieldServiceAccountName          = "service_account_name"
+	fieldKubernetesRoleName          = "kubernetes_role_name"
+	fieldKubernetesRoleType          = "kubernetes_role_type"
+	fieldGeneratedRoleRules          = "generated_role_rules"
+	fieldNameTemplate                = "name_template"
+	fieldExtraAnnotations            = "extra_annotations"
+	fieldExtraLabels                 = "extra_labels"
+)
+
+func kubernetesSecretBackendRoleResource() *schema.Resource {
+	return &schema.Resource{
+		CreateContext: kubernetesSecretBackendRoleCreateUpdate,
+		ReadContext:   kubernetesSecretBackendRoleRead,
+		UpdateContext: kubernetesSecretBackendRoleCreateUpdate,
+		DeleteContext: kubernetesSecretBackendRoleDelete,
+		Importer: &schema.ResourceImporter{
+			StateContext: schema.ImportStatePassthroughContext,
+		},
+		Schema: map[string]*schema.Schema{
+			consts.FieldName: {
+				Type:        schema.TypeString,
+				Description: "The name of the role.",
+				ForceNew:    true,
+				Required:    true,
+			},
+			consts.FieldBackend: {
+				Type:        schema.TypeString,
+				Description: "The mount path for the Kubernetes secrets engine.",
+				Required:    true,
+				ForceNew:    true,
+				StateFunc: func(v interface{}) string {
+					return strings.Trim(v.(string), "/")
+				},
+			},
+			fieldAllowedKubernetesNamespaces: {
+				Type: schema.TypeList,
+				Elem: &schema.Schema{
+					Type: schema.TypeString,
+				},
+				Description: "The list of Kubernetes namespaces this role can generate " +
+					"credentials for. If set to '*' all namespaces are allowed.",
+				Required: true,
+			},
+			fieldTokenMaxTTL: {
+				Type: schema.TypeString,
```

Comment:
I thought that we could take in Go duration strings as config, but it turns out this is a bit difficult to get right. The API takes a Go duration string and returns integer seconds. I thought I could use [`flattenVaultDuration`](https://github.com/hashicorp/terraform-provider-vault/blob/main/vault/structures.go#L113) to achieve going back from integer seconds to duration string, but it doesn't work correctly in some situations.

Is anyone aware of examples of how to do this properly? If not, I'll change these to take in integer seconds instead.

Reply:
Hmm, I tried to play around with some different options here and didn't find anything I liked. I'm going to switch these over to seconds and add a `_seconds` suffix to the fields. Thanks, all!

Reply:
I think it is fine to forgo the extra `_seconds` suffix. We have standardized on seconds (int) for all TTL type fields, since it better matches the read from Vault.

Reply:
Changed this to use integer seconds in cc3b1aa. I left the `_seconds` suffix off of the field name.